src/devices/machine/w5100.cpp

```diff
@@ -0,0 +1,4320 @@
+}
+
+
+
+/*
+INTPTMR sets internal Interrupt Pending Timer Count.
+
+When INTn is de-asserted to High, Timer Count is initialized to INTPTMR and decreased by 1 from initial value to ‘0’ every SYS_CLK x 4.
+When Interrupt occurs and the corresponding Interrupt Mask is set and INTPTMR is ‘0’, INTn is asserted to Low.
+
+INTn is pin 47,
+
+See How To Use Interrupt Application Note.
+*/
+
+
+TIMER_CALLBACK_MEMBER(w5100_base_device::interrupt_timer)
+{
+	// w5100s delayed IRQ timer
+	if (m_irq_state) m_irq_handler(ASSERT_LINE);
+}
+
+
+void w5100_base_device::update_ethernet_irq()
+{
+	int ir = m_common_registers[IR] & 0b11100000;
+	int ir2 = 0;
+	int slir = 0;
+
+	for (int sn = 0, bit = IR_S0_INT; sn < 4; ++sn, bit <<= 1)
+	{
+		const uint8_t *socket = m_socket_registers[sn];
+		int sir = socket[Sn_IR] & 0b00011111;
+
+		if (m_device_type == dev_type::W5100S)
+			sir &= socket[Sn_IMR];
+
+		if (sir) ir |= bit;
+	}
+
+	m_common_registers[IR] = ir;
+	ir &= m_common_registers[IMR];
+
+	int intptmr = 0;
+
+	if (m_device_type == dev_type::W5100S)
+	{
+		ir2 = m_common_registers[IR2] & 0b00000001;
+		ir2 &= m_common_registers[IMR2];
+		slir = m_common_registers[SLIR] & 0b00000111;
+		slir &= m_common_registers[SLIMR];
+
+		if ((m_common_registers[MR2] & MR2_IEN) == 0)
+		{
+			ir = 0;
+			ir2 = 0;
+			slir = 0;
+		}
+		intptmr = read16(m_common_registers + INTPTMR0);
+	}
+
+	uint32_t new_state = (slir << 16) | (ir2 << 8) | (ir);
+
+
+	if (new_state ^ m_irq_state)
+	{
+		m_irq_state = new_state;
+
+		if (intptmr)
+		{
+			// w5100s only.
+			// pending interrupt timer is kicked off when interrupts are cleared.
+			// while the timer is active, interrupts will be defered.
+			if (m_irq_state && m_interrupt_timer->enabled())
+				return;
+			if (!m_irq_state && !m_interrupt_timer->enabled())
+					m_interrupt_timer->adjust(clocks_to_attotime(intptmr * 4));
+		}
+
+		m_irq_handler(m_irq_state ? ASSERT_LINE : CLEAR_LINE);
+	}
+
+}
+
+int w5100_base_device::register_bank(uint16_t offset)
+{
+	if ((offset & 0x0700) < 0x0400) return 0;
+
+	if (m_device_type == dev_type::W5100)
+	{
+		if (offset > 0x07ff) return 7;
+	}
+	return (offset & 0x0700) >> 8;
+}
+
+
+/*
+ * Direct bus interface: 15-bit address, 8-bit data bus
+ * Indirect bus interface: 2-bit address, 8-bit data bus
+ * W5100s is always indirect w/ auto-increment
+ */
+
+uint8_t w5100_base_device::read(uint16_t offset)
+{
+	if (m_common_registers[0] & MR_IND)
+	{
+		switch(offset & 0x03)
+		{
+			case IDM_OR:
+				return m_common_registers[0];
+				break;
+			case IDM_AR0:
+				return m_idm >> 8;
+				break;
+			case IDM_AR1:
+				return m_idm & 0xff;
+				break;
+			case IDM_DR:
+				offset = m_idm;
+
+				if ((m_common_registers[0] & MR_AI) && !machine().side_effects_disabled())
+				{
+					m_idm++;
+					if (m_device_type == dev_type::W5100 && (m_idm & 0x1fff) == 0x0000)
+						m_idm -= 0x2000;
+				}
+				break;
+		}
+	}
+
+	offset &= 0x7fff;
+
+	return m_address_space->read_byte(offset);
```

Comment:
Using an `address_space` directly is inefficient – you usually want to use an `address_space_specific` or `address_space_cache`.